File: backend/src/auto_approps/doc_parser.py
from __future__ import annotations

import io
from docx import Document
from docx.oxml.ns import qn
from docx.table import Table
from docx.text.paragraph import Paragraph

from .models import DocChunk, ParsedDocument


def iter_block_items(parent):
    """Yield paragraphs and tables in document order."""
    body = parent.element.body
    for child in body.iterchildren():
        if child.tag == qn("w:p"):
            yield Paragraph(child, parent)
        elif child.tag == qn("w:tbl"):
            yield Table(child, parent)


def parse_docx(file_bytes: bytes, filename: str) -> ParsedDocument:
    doc = Document(io.BytesIO(file_bytes))
    chunks: list[DocChunk] = []
    full_text_parts: list[str] = []
    current_heading = ""
    chunk_index = 0

    for block in iter_block_items(doc):
        if isinstance(block, Paragraph):
            text = block.text.strip()
            if not text:
                continue

            style_name = (block.style.name or "").lower()
            if "heading" in style_name:
                current_heading = text
                source = f"Heading: '{text}'"
                chunk_type = "heading"
                # Extract heading level: "heading 1" -> 1, "heading 2" -> 2, default 2
                heading_level = 2
                for part in style_name.split():
                    if part.isdigit():
                        heading_level = int(part)
                        break
                full_text_parts.append(f"\n## {text}\n")
            else:
                source = f"Section '{current_heading}' > Paragraph {chunk_index + 1}" if current_heading else f"Paragraph {chunk_index + 1}"
                chunk_type = "paragraph"
                full_text_parts.append(text)

            chunks.append(DocChunk(
                text=text,
                source_location=source,
                chunk_type=chunk_type,
                heading_context=current_heading,
                heading_level=heading_level if chunk_type == "heading" else 0,
                index=chunk_index,
            ))
            chunk_index += 1

        elif isinstance(block, Table):
            headers = [cell.text.strip() for cell in block.rows[0].cells] if block.rows else []

            for row_idx, row in enumerate(block.rows):
                cells = [cell.text.strip() for cell in row.cells]
                if row_idx == 0:
                    # Header row
                    row_text = " | ".join(cells)
                    full_text_parts.append(row_text)
                    full_text_parts.append("-" * len(row_text))
                else:
                    if headers:
                        pairs = [f"{h}: {c}" for h, c in zip(headers, cells) if c]
                        row_text = " | ".join(pairs)
                    else:
                        row_text = " | ".join(cells)
                    full_text_parts.append(row_text)

                source = f"Section '{current_heading}' > Table Row {row_idx + 1}" if current_heading else f"Table Row {row_idx + 1}"
                chunks.append(DocChunk(
                    text=row_text if row_idx > 0 else " | ".join(cells),
                    source_location=source,
                    chunk_type="table_row",
                    heading_context=current_heading,
                    index=chunk_index,
                ))
                chunk_index += 1

    return ParsedDocument(
        filename=filename,
        chunks=chunks,
        full_text="\n".join(full_text_parts),
    )

File: electron/services/doc-parser.ts
import mammoth from "mammoth";
import * as cheerio from "cheerio";
import type { DocChunk, ParsedDocument } from "./models.js";

/**
 * Parse a .docx file buffer into a structured ParsedDocument.
 *
 * Uses mammoth to convert the docx to HTML, then cheerio to walk the HTML
 * in document order:
 *   - Headings (h1-h6) become "heading" chunks
 *   - Paragraphs (p)   become "paragraph" chunks
 *   - Table rows (tr)  become "table_row" chunks, with the first row treated
 *     as headers and subsequent rows formatted as "header: cell" pairs.
 */
export async function parseDocx(
  fileBuffer: Buffer,
  filename: string,
): Promise<ParsedDocument> {
  const result = await mammoth.convertToHtml({ buffer: fileBuffer });
  const $ = cheerio.load(result.value);

  const chunks: DocChunk[] = [];
  const fullTextParts: string[] = [];
  let currentHeading = "";
  let chunkIndex = 0;

  // Walk top-level body children in document order so headings, paragraphs,
  // and tables are encountered in their original sequence.
  $("body")
    .children()
    .each((_i, el) => {
      const tagName = ((el as any).tagName ?? (el as any).name ?? "").toLowerCase();

      // ── Headings ──────────────────────────────────────────────────────
      const headingMatch = tagName.match(/^h([1-6])$/);
      if (headingMatch) {
        const text = $(el).text().trim();
        if (!text) return;

        const headingLevel = parseInt(headingMatch[1], 10);
        currentHeading = text;

        const sourceLocation = `Heading: '${text}'`;
        fullTextParts.push(`\n## ${text}\n`);

        chunks.push({
          text,
          source_location: sourceLocation,
          chunk_type: "heading",
          heading_context: currentHeading,
          heading_level: headingLevel,
          index: chunkIndex,
        });
        chunkIndex++;
        return;
      }

      // ── Paragraphs ───────────────────────────────────────────────────
      if (tagName === "p") {
        const text = $(el).text().trim();
        if (!text) return;

        const sourceLocation = currentHeading
          ? `Section '${currentHeading}' > Paragraph ${chunkIndex + 1}`
          : `Paragraph ${chunkIndex + 1}`;

        fullTextParts.push(text);

        chunks.push({
          text,
          source_location: sourceLocation,
          chunk_type: "paragraph",
          heading_context: currentHeading,
          heading_level: 0,
          index: chunkIndex,
        });
        chunkIndex++;
        return;
      }

      // ── Tables ────────────────────────────────────────────────────────
      if (tagName === "table") {
        const rows = $(el).find("tr");
        if (rows.length === 0) return;

        // Extract header cells from the first row.
        const headers: string[] = [];
        rows
          .first()
          .find("td, th")
          .each((_j, cell) => {
            headers.push($(cell).text().trim());
          });

        rows.each((rowIdx, row) => {
          const cells: string[] = [];
          $(row)
            .find("td, th")
            .each((_j, cell) => {
              cells.push($(cell).text().trim());
            });

          let rowText: string;

          if (rowIdx === 0) {
            // Header row – plain pipe-delimited
            rowText = cells.join(" | ");
            fullTextParts.push(rowText);
            fullTextParts.push("-".repeat(rowText.length));
          } else {
            // Data rows – pair with headers when available
            if (headers.length > 0) {
              const pairs: string[] = [];
              for (let k = 0; k < cells.length; k++) {
                if (cells[k]) {
                  const header = k < headers.length ? headers[k] : `Col${k + 1}`;
                  pairs.push(`${header}: ${cells[k]}`);
                }
              }
              rowText = pairs.join(" | ");
            } else {
              rowText = cells.join(" | ");
            }
            fullTextParts.push(rowText);
          }

          const sourceLocation = currentHeading
            ? `Section '${currentHeading}' > Table Row ${rowIdx + 1}`
            : `Table Row ${rowIdx + 1}`;

          chunks.push({
            text: rowIdx === 0 ? cells.join(" | ") : rowText,
            source_location: sourceLocation,
            chunk_type: "table_row",
            heading_context: currentHeading,
            heading_level: 0,
            index: chunkIndex,
          });
          chunkIndex++;
        });
      }
    });

  return {
    filename,
    chunks,
    full_text: fullTextParts.join("\n"),
  };
}
